python/lsst/daf/butler/script/ingest_files.py

```diff
@@ -0,0 +1,140 @@
+                 formatter: Optional[str] = None,
+                 id_generation_mode: str = "UNIQUE",
+                 prefix: Optional[str] = None,
+                 transfer: str = "auto") -> None:
+    """Ingest files from a table.
+
+    Parameters
+    ----------
+    repo : `str`
+        URI string of the Butler repo to use.
+    dataset_type : `str`
+        The name of the dataset type for the files to be ingested.
+    formatter : `str`
+        Fully-qualified python class name for the `Formatter` to use
+        to read the ingested files.
+    run : `str`
+        The run in which the files should be ingested.
+    table_file : `str`
+        Path to a table file to read.
+    id_generation_mode : `str`
+        Mode to use for generating IDs.  Should map to `DatasetGenIdEnum`.
+    prefix : `str`
+        Prefix to use when resolving relative paths in table files. The default
+        is to use the current working directory.
+    transfer : `str`
+        Transfer mode to use for ingest.
+    """
+
+    # Check that the formatter can be imported -- validate this as soon
+    # as possible before we read a potentially large table file.
+    if formatter:
+        doImport(formatter)
+    else:
+        formatter = None
+
+    # Force null prefix to None for API compatibility.
+    if not prefix:
+        prefix = None
+
+    # Convert the dataset ID gen more string to enum
+    id_gen_mode = DatasetIdGenEnum.__members__[id_generation_mode]
+
+    # Create the butler with the relevant run attached.
+    butler = Butler(repo, run=run)
+
+    datasetType = butler.registry.getDatasetType(dataset_type)
+
+    universe = butler.registry.dimensions
+
+    # Convert any additional k=v strings in the dataId tuple to dict
+    # form.
+    common_data_id: Dict[str, Any] = {}
+    for id_str in data_id:
+        dimension_str, value = id_str.split("=")
+
+        try:
+            dimension = universe.getStaticDimensions()[dimension_str]
+        except KeyError:
+            raise ValueError(f"DataID dimension '{dimension_str}' is not known to this universe.") from None
+
+        # Cast the value to the right python type (since they will be
+        # strings at this point).
+        value = dimension.primaryKey.getPythonType()(value)
+
+        common_data_id[dimension_str] = value
+
+    # Read the table and determine the dimensions (the first column
+    # is the file URI).
+    table = Table.read(table_file)
+
+    dimensions = table.colnames
+    dimensions.pop(0)
+
+    # Handle multiple dataIds per file by grouping by file.
+    refs_by_file = defaultdict(list)
+    n_datasets = 0
+    for row in table:
+        dataId: Dict[str, Any] = {k: row[k] for k in dimensions}
+
+        # The command line can override a column.
+        dataId.update(common_data_id)
+
+        standardized = DataCoordinate.standardize(dataId, graph=datasetType.dimensions)
+
+        ref = DatasetRef(datasetType, standardized, conform=False)
```

Comment:
Funny you should ask and maybe @TallJimbo can explain. The reason I did it this way is because of mypy which insists that DatasetRef can only take a DataCoordinate even though it can also take a dict if conform=True. Maybe I should have asked @TallJimbo if I can fix the annotation for the DatasetRef constructor itself.

Reply:
Yeah, sounds like the annotation should probably be changed to the `DataId` type alias.

Reply:
I changed the annotation and everything broke since the overwhelming usage for `conform=False` is for plain `dict` in test and validation code and to appease mypy I had made that illegal. There doesn't seem to be a way to turn a non-conforming `dict` into a non-conforming `DataCoordinate` so I think I'm going to have to tell `mypy` that the `self.dataId` itself can also be `DataId` and is not required to be a `DataCoordinate`.

Reply:
...and to confirm, mypy is really upset if I declare that `DatasetRef.dataId` might not be a `DataCoordinate`. This all explains why we weren't telling mypy about the plain dict option. I'll go back to how it was I think and add an `# type: ignore` (which is what we do in `_butler.py`.